### ISSUE TITLE
[c++] syntax highlighting incorrect due to macro

### ISSUE THREAD

**Issue by wjwwood** [sublimehq/Packages#452]
I'm using Sublime Text 3 build 3114:

<img width="437" alt="screen shot 2016-06-03 at 11 02 47 am" src="https://cloud.githubusercontent.com/assets/100427/15788125/b9f3a954-297a-11e6-9cba-702e0ff671f6.png">

For example:

<img width="1388" alt="screen shot 2016-06-03 at 10 55 21 am" src="https://cloud.githubusercontent.com/assets/100427/15787976/02293afa-297a-11e6-9fa4-52879810b966.png">

If I remove or comment out the macro, the syntax highlighting is fixed until the next macro:

<img width="1388" alt="screen shot 2016-06-03 at 10 58 38 am" src="https://cloud.githubusercontent.com/assets/100427/15787994/24c1fd68-297a-11e6-8097-30b290ca5fcd.png">

You can see the the next function declaration (after the one with the commented macro) has the highlighting swapped on the function name and the return type. If I comment the macro on the function with the swapped highlighting it moves to the next one:

<img width="1388" alt="screen shot 2016-06-03 at 11 00 18 am" src="https://cloud.githubusercontent.com/assets/100427/15788047/6086aae2-297a-11e6-912f-98b738c08303.png">

Here is the exact version of the file in question:

https://github.com/ros2/rclcpp/blob/6adfb917a9ddf44e240c44955accd60c68ec3802/rclcpp/include/rclcpp/node.hpp#L221

It highlighted correctly with a recent ST3 build, but I can't say exactly which one. It compiles correctly and GitHub's web view manages to highlight it correctly too, so I think it's valid code. It seems to be an interaction between the macro and the somewhat complex templated return types or parameters possibly. I'm not sure how to debug this issue, but if I can give more info that would be helpful I'll try to do so.

I looked for a similar issue here and on the other issue tracker and found this issue:

https://github.com/SublimeTextIssues/DefaultPackages/issues/119

But it seems to have been closed as fixed for build 3112, which is older than what I have. I didn't see anything else that seemed to be a duplicate. Sorry if I missed something.

**Comment by wjwwood** [comment]
Another example, this one with failing to match closing `)` and `}`:

File https://github.com/ros2/rclcpp/blob/2be9568498a82c3c28805039c1037c0427cea1ae/rclcpp/include/rclcpp/parameter_client.hpp#L156:
<img width="1388" alt="screen shot 2016-06-03 at 12 08 12 pm" src="https://cloud.githubusercontent.com/assets/100427/15790037/06f3f20a-2984-11e6-855b-0f18b30bf447.png">

**Comment by wbond** [comment]
The issue is related to linebreaks and how C/C++, combined with the preprocessor allow for all sorts of bonkers formatting that can be hard to parse.

I will see if I can adjust the heuristics to get this code to highlight properly.

**Comment by wjwwood** [comment]
@wbond thanks! let me know if I can help with testing.

**Comment by FichteFoll** [comment]
Looks like the same issue as https://github.com/sublimehq/Packages/issues/433 to me.

**Comment by wjwwood** [comment]
@FichteFoll looks like it, sorry I was looking for C++ only issues. Is there a good way to try out a change? Maybe a howto on what packages to pull or patch to try it out?

**Comment by wjwwood** [comment]
I tried applying https://github.com/sublimehq/Packages/commit/6a5e2d46d01df3c5da0d3cd181ea62fffdc7aa76.diff to my local copy of `C++.sublime-package` by unzipping it, patching the files, and then zipping it back, but that doesn't see to have worked. Is there a better way to test this out?

**Comment by wbond** [comment]
I tried changing the function definition to all on one line, but that didn't help. I believe it is related to generics syntax.

In terms of trying out the latest version, using the instructions in the readme of this repo will allow you to use it. You could then use git to checkout a specific commit if you want.

**Comment by wbond** [comment]
This should be fixed by 8dc15562acc66f631c22856cbc798740bb36f165

**Comment by wjwwood** [comment]
Awesome! Thanks, I'll give it a shot.

**Comment by wjwwood** [comment]
After a quick test, it looks like it's working well.

BTW, the instructions in the README will only install (by symlink) the Python package from this directory, and not, for example, the C++ package:

https://github.com/sublimehq/Packages#os-x-and-linux

It might be good to make a note of that in the README.

Thanks again.
